### ISSUE TITLE
[SOS] Intermittent Connection Issues with HttpClient in .NET 6

### ISSUE THREAD

**Issue by yuefengkai** [dotnet/runtime#101407]
Hello,

I am experiencing intermittent connectivity issues when making requests to https://api.maersk.com/ using HttpClient in .NET 6. While I can occasionally establish a connection, most attempts fail. Interestingly, when I make the same requests using Python code, the connection is consistently successful.

Here are some troubleshooting steps I have already tried:

Modifying the HTTP protocol version.
Changing HTTP headers.
Confirmed that I can successfully ping and curl the URL from my location in Hong Kong.
Despite these efforts, the issue persists specifically with C# HttpClient. Below is the code snippet I am using:

Could you help me identify what might be causing this inconsistency?

Thank you for your assistance.
![image](https://github.com/microsoft/dotnet/assets/2100585/87f40ff5-1910-4bc3-89ff-65c269e4ce1c)
![image](https://github.com/microsoft/dotnet/assets/2100585/70afc7be-2772-4f9c-b000-9a093bb71402)

```
 using System.Net;
using System.Diagnostics;
using System.Net.Http;

Console.WriteLine("Hello, World!");


var httpclietHandel = new HttpClientHandler
{

        AutomaticDecompression = System.Net.DecompressionMethods.GZip,
        ServerCertificateCustomValidationCallback = (message, cert, chain, errors) => { return true; },
         SslProtocols = System.Security.Authentication.SslProtocols.Tls13
};
var client = new HttpClient(httpclietHandel)
{
        //DefaultRequestVersion = HttpVersion.Version11,
        //  DefaultVersionPolicy = HttpVersionPolicy.RequestVersionOrLower
};
var request = new HttpRequestMessage
{
        Method = HttpMethod.Get,
        RequestUri = new Uri("https://api.maersk.com/"),
        Headers =
                                {
                                                                { "Consumer-Key", "xxxxxxxxxx" },
                                                                {"Accept","text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7"},
                                                                {"Accept-Encoding","gzip, deflate, br, zstd"},
                                                                {"User-Agent","Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36"},
                                                                {"Accept-Language","zh-CN,zh-TW;q=0.9,zh;q=0.8,en;q=0.7"},
                },
};

Stopwatch sw = Stopwatch.StartNew();

using (var response = await client.SendAsync(request))
{
        response.EnsureSuccessStatusCode();
        var body = await response.Content.ReadAsStringAsync();

        sw.Stop();
        Console.WriteLine($"{sw.ElapsedMilliseconds}-=--{body}");
}

Console.WriteLine("Done");
```

```
import requests
import time

payload = "{}"

headers = {
    'Consumer-Key': "xxxxxx",
    'content-type': "application/json"
    }

url = "https://api.maersk.com/"

start = time.time()
response = requests.get(url,timeout=3,headers=headers,data=payload,)

print(response.status_code)
print(response.content)
print("Finished...{}".format(time.time()-start))
```

**Comment by huoyaoyuan** [comment]
I can't reproduce your issue on Windows machine. Is it TLS1.3 related? Can you provide more information about the machine you are running the code?

**Comment by wfurt** [comment]
You should get packet capture @yuefengkai. That should provide some insight. You can try to flip `AutomaticDecompression`  but I doubt that would be the root cause.

**Comment by yuefengkai** [comment]
> I can't reproduce your issue on Windows machine. Is it TLS1.3 related? Can you provide more information about the machine you are running the code?我无法在 Windows 计算机上重现您的问题。与 TLS1.3 相关吗？您能否提供有关运行代码的机器的更多信息？

I have deployed on Linux and tried various protocols including HTTP/1.0, HTTP/1.1, HTTP/2.0, and HTTP/3.0, but none have worked.

**Comment by yuefengkai** [comment]
> You should get packet capture @yuefengkai. That should provide some insight. You can try to flip `AutomaticDecompression` but I doubt that would be the root cause.你应该获得数据包捕获。这应该提供一些见解。您可以尝试翻转 `AutomaticDecompression` 但我怀疑这将是根本原因。

How should I capture packets? Should I use tcpdump?

**Comment by huoyaoyuan** [comment]
> tried various protocols including HTTP/1.0, HTTP/1.1, HTTP/2.0, and HTTP/3.0

I mean this:
```csharp
SslProtocols = System.Security.Authentication.SslProtocols.Tls13
```

Providing more information about your Linux distro may be helpful.

**Comment by wfurt** [comment]
yes, tcpdump would work. Or more conveniently Wireshark if you have GUI. From the exception it seems like handshake already succeeded.

**Comment by yuefengkai** [comment]
> Linux

![image](https://github.com/dotnet/runtime/assets/2100585/d14327e8-a477-4dd5-ae38-db481e48c28c)

**Comment by yuefengkai** [comment]
> yes, tcpdump would work. Or more conveniently Wireshark if you have GUI. From the exception it seems like handshake already succeeded.

[packets_capture.zip](https://github.com/dotnet/runtime/files/15071868/packets_capture.zip)

**Comment by wfurt** [comment]
You can see that client sends data in packet 13 and never gets answer back from the server. Figuring out why may be difficult. If the python fragment works I would start looking at differences in the HTTP request. Some severs are very picky about headers and they can fail in unusual ways.  And the clients may have different default sets so make sure they are 100% identical.

You can also experiment with `openssl s_client -connect xxx:xx`. That would allow you to open TLS to the server and then you can cut & paste the HTTP/1.1 request to it to see if that yields some more insight.  

It seems like this issue is specific to your environment so I think it needs to be investigated there.

**Comment by yuefengkai** [comment]
> api.maersk.com/

```
openssl s_client -connect api.maersk.com:443
CONNECTED(00000003)
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
verify return:1
depth=0 CN = api.maersk.com
verify return:1
---
Certificate chain
 0 s:CN = api.maersk.com
   i:C = US, O = Let's Encrypt, CN = R3
 1 s:C = US, O = Let's Encrypt, CN = R3
   i:C = US, O = Internet Security Research Group, CN = ISRG Root X1
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIEHTCCAwWgAwIBAgISBBZXBMR3YFVvPNsarJms6KnYMA0GCSqGSIb3DQEBCwUA
MDIxCzAJBgNVBAYTAlVTMRYwFAYDVQQKEw1MZXQncyBFbmNyeXB0MQswCQYDVQQD
EwJSMzAeFw0yNDA0MDQxMTM2MDdaFw0yNDA3MDMxMTM2MDZaMBkxFzAVBgNVBAMT
DmFwaS5tYWVyc2suY29tMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE506Xy+dJ
oQ4oBmL6p3ywXa0pchyCSPdu5HgHplbD2jdLC4+jZGQloPPDAXuEZz8nSOABo21m
dStV7+UnXB6rkqOCAg8wggILMA4GA1UdDwEB/wQEAwIHgDAdBgNVHSUEFjAUBggr
BgEFBQcDAQYIKwYBBQUHAwIwDAYDVR0TAQH/BAIwADAdBgNVHQ4EFgQUTt/vH/Sq
SRGZCDhz0+jWCFhgT4kwHwYDVR0jBBgwFoAUFC6zF7dYVsuuUAlA5h+vnYsUwsYw
VQYIKwYBBQUHAQEESTBHMCEGCCsGAQUFBzABhhVodHRwOi8vcjMuby5sZW5jci5v
cmcwIgYIKwYBBQUHMAKGFmh0dHA6Ly9yMy5pLmxlbmNyLm9yZy8wGQYDVR0RBBIw
EIIOYXBpLm1hZXJzay5jb20wEwYDVR0gBAwwCjAIBgZngQwBAgEwggEDBgorBgEE
AdZ5AgQCBIH0BIHxAO8AdgBIsONr2qZHNA/lagL6nTDrHFIBy1bdLIHZu7+rOdiE
cwAAAY6pG/xfAAAEAwBHMEUCIQDvWgbqW3/t3bJ11yD3iej2zfD80Q94uOlck5PT
xvXxqgIgWwxI1gTs11cg3rUZ6ILrFrHcZOG0SkUWkJ6DRinFbZMAdQDuzdBk1dsa
zsVct520zROiModGfLzs3sNRSFlGcR+1mwAAAY6pG/xkAAAEAwBGMEQCIFJxSBlw
hOUnqTFvmOFSCQyvppxvyRBOA6iw5KVroC+VAiAWOIim0lBfWjp4efpfaXY0CiGj
kF/jzccHJRmtYrtmaTANBgkqhkiG9w0BAQsFAAOCAQEAK1qsQqQXGqxmbUZ2bOkJ
VQzjBbwYZ4+ly+RMSAH1dY+B7N3nLSRUaWRPls+ztreD94mq28prA7tiIeVC/FBu
bbr05hTiKgll8XdiBkEuK6Xp2LdpQql1jt+NdK0m5WT8WcVxfhA/U2BmpJHfpgmV
OAiOF0REI8jDsmqTv1HmIVJGgRH6xBm1IPcXRYG2GTUavISHEaAiS5pAMRS1Qenl
dXA+u68/ThEYaPsghkYZtiDlYEwdTOfB5/IeXHNXbNGs31im3EagXQithq+twvWr
6zjeZOXMC6kzIm/GUjlHuzuIcMiGDvaxlLdpqY4D0bO7DFV4HPjxl32X65wraVg4
lA==
-----END CERTIFICATE-----
subject=CN = api.maersk.com

issuer=C = US, O = Let's Encrypt, CN = R3

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: ECDSA
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2754 bytes and written 392 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 256 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: F7D6A0753598E1BE62ED3C53ABD767CAD73514D1AD5951F6081AF3CEDCB9FD22
    Session-ID-ctx: 
    Resumption PSK: BC6E2AC3471AED54174C56F16846A731D399C4514012F188ADA026EE3628B0299C35BEAB9E66CE6BEE41B91E5C6E1077
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 83100 (seconds)
    TLS session ticket:
    0000 - 00 03 26 c2 db ef ee e7-55 1f 19 bb 87 67 5d 18   ..&.....U....g].
    0010 - 88 d0 7f c9 82 14 c3 85-52 17 90 ca 20 94 36 04   ........R... .6.
    0020 - ed 81 0c 20 5b ed 2a 7f-53 6c a7 8b 44 0f d2 4b   ... [.*.Sl..D..K
    0030 - 9b d1 3b 10 7f 03 87 8c-bf 7e b7 e3 d8 27 4c 46   ..;......~...'LF
    0040 - 33 f4 d9 d0 7c aa 05 90-89 e4 b4 f5 3f 00 2b d0   3...|.......?.+.
    0050 - b8 94 89 d9 a3 2f 99 01-f9 d0 66 98 a0 74 62 e9   ...../....f..tb.
    0060 - ef cc b7 3a f8 e0 bb 9b-b7 48 65 09 12 e8 70 e0   ...:.....He...p.
    0070 - ce bd 1d f1 40 31 7e 58-00 3d 4d fe 07 ac e3 dd   ....@1~X.=M.....
    0080 - 98 c6 cb 46 97 10 4c 6c-b6 0a b1 9e 9f be 7c 60   ...F..Ll......|`
    0090 - e2 81 bf 71 fe 7d e0 2f-bd 87 d7 81 42 83 2a 7b   ...q.}./....B.*{
    00a0 - 2d 7a e5 8f b8 47 05 23-3b 16 9d f5 93 b8 c2 24   -z...G.#;......$
    00b0 - eb 7a 82 4d 2d ab 48 e4-11 99 80 0b 69 98 94 28   .z.M-.H.....i..(
    00c0 - 61 c3 0f dc 0b 47 c7 a4-67 4f 6e 86 57 cf 9d 56   a....G..gOn.W..V
    00d0 - 29 d2 d8 8d a3 eb 6b 4a-b9 8e f1 87 0d 13 2a 70   ).....kJ......*p

    Start Time: 1713923178
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 4022815410400E554E8FF1E3A509ACFD7590310C85E04DE99ED63EC467F38C39
    Session-ID-ctx: 
    Resumption PSK: F8650D74BEC27967888069E2093B247496AE2831170D544E2619166A51EB42740738A70D5338DB58C30D029C442D7A3C
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 83100 (seconds)
    TLS session ticket:
    0000 - 00 03 26 c2 db ef ee e7-55 1f 19 bb 87 67 5d 18   ..&.....U....g].
    0010 - 87 f2 4f c0 83 0b 56 5b-f6 9e 34 55 0d 34 51 2b   ..O...V[..4U.4Q+
    0020 - e1 75 8f 06 16 80 00 24-a9 4a 99 1d 32 ff 8e 68   .u.....$.J..2..h
    0030 - 08 44 9f bc 75 e2 70 9b-b1 70 91 03 a1 55 4b d0   .D..u.p..p...UK.
    0040 - 52 c2 16 c6 0d 30 36 b2-14 1d 17 ba 54 e5 ba 18   R....06.....T...
    0050 - a0 4d 16 27 2b 39 e4 37-58 3a 26 b6 23 d2 d6 96   .M.'+9.7X:&.#...
    0060 - 83 ae 3e 6a 26 be e1 69-2d a8 07 3e c7 be 91 97   ..>j&..i-..>....
    0070 - 53 b7 64 2f e9 76 3f 18-0f 9f 32 4f 71 f0 b1 66   S.d/.v?...2Oq..f
    0080 - f0 55 4e 2c 1d 6a f9 93-63 2e ab 76 c3 d9 40 dd   .UN,.j..c..v..@.
    0090 - 2b 17 c8 88 b8 b6 44 46-ca f1 f4 1c 64 e4 60 9b   +.....DF....d.`.
    00a0 - 8e ca f6 ba 2c 0e 1f b2-18 c4 fb 5c 54 06 2d bd   ....,......\T.-.
    00b0 - 12 bf 69 3f e9 63 27 43-94 ef 08 d0 fe c5 d1 85   ..i?.c'C........
    00c0 - 5f e1 22 ea ac d4 ab a2-b0 31 ec 62 4c dc 2c a0   _."......1.bL.,.
    00d0 - ab 3c 16 a0 ab e7 6d 5d-da c6 31 05 6d 6a 1a 73   .<....m]..1.mj.s

    Start Time: 1713923178
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK


closed
```

**Comment by CarnaViire** [comment]
@yuefengkai after the ssl connection gets established, but before you get `closed`, you can cut & paste the HTTP/1.1 request as @wfurt mentioned. E.g.:

```diff
        ....
        00c0 - c2 05 50 54 bd 7a 9a ff-9d b1 35 df 5d d7 c4 cb   ..PT.z....5.]...
        00d0 - 86 bc 70 68 4b d2 1a 4d-6d b5 00 8f d9 b7 2a 5a   ..phK..Mm.....*Z

        Start Time: 1714403582
        Timeout   : 7200 (sec)
        Verify return code: 0 (ok)
        Extended master secret: no
        Max Early Data: 0
    ---
    read R BLOCK
+   GET / HTTP/1.1
+   Host: api.maersk.com
+   Consumer-Key: xxxxxxxxxx
+   Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
+   Accept-Encoding: gzip, deflate, br, zstd
+   ser-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36
+   Accept-Language: zh-CN,zh-TW;q=0.9,zh;q=0.8,en;q=0.7
+
    HTTP/1.1 401 Unauthorized
    Content-Type: application/json
    Content-Length: 146
    Edge-Cache-Tag: api.maersk.com/
    Date: Mon, 29 Apr 2024 15:13:09 GMT
    ....
    Set-Cookie: ....; Domain=.maersk.com; Path=/; Expires=Mon, 29 Apr 2024 19:13:09 GMT; Max-Age=14400

    {"error":[{"code":"ERR_GW_001","message":"API Key Validation Failed, please check the API key","reason":"Invalid or expired Key","status":"401"}]}
```

BTW I cannot reproduce it even on your distro. Dockerfile I was using:

<details>

```dockerfile
FROM alibaba-cloud-linux-3-registry.cn-hangzhou.cr.aliyuncs.com/alinux3/alinux3:latest

RUN yum install -y dnf

RUN dnf install -y \
        dnf-plugins-core \
        libicu \
        vim

RUN dnf install -y \
        tar \
        findutils \
        openssl

RUN curl -sSL https://dot.net/v1/dotnet-install.sh | bash /dev/stdin --channel 6.0.1xx --install-dir /usr/share/dotnet 

COPY . ./Test
WORKDIR /Test
CMD /usr/share/dotnet/dotnet run -c release
```

'+ Test.csproj with 6.0 TFM
'+ Program.cs containing the code from the top post 

</details>

The output gives me the expected 401

<details>

```
➜ alinux  ✗ docker run -it --rm alinux-3-test

Welcome to .NET 6.0!
---------------------
SDK Version: 6.0.129

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
Hello, World!
Unhandled exception. System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at Program.<Main>$(String[] args) in /Test/Program.cs:line 38
   at Program.<Main>(String[] args)
```

</details>
